### PR TITLE
Improving prickup: prioritize hovered blobs 

### PR DESCRIPTION
**Pull request by Nananas**
## Status

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description

When hovering the a blob, prioritize pickup of this blob. Checks which blob is drawn at top if stacked.
Related video: https://discord.com/channels/337250580303052801/337266863841214465/756102703280553994

This PR currently uses the CShape of an item to check mouse overlap, together with the draw order. 
However, some items such as bomb-arrows, gold and the flag have CShapes that differ from the actual sprite.
Modifying the CShape to match the sprite does not feel like the correct solution.

After testing, I found some issues with the a previous engine change, which should be fixed in Engine PR `#204`.

Commit 23271f9 adds the z-ordering (SetZ) and "hover-polygons" of some "important" items to draw them on top of other stuff. This is done by adding the ImportantPickup.as script, also see the comments in that file. 

I already added "hover-polygons" to the flag and the special arrows, as they differ a lot from their CShape.

(This PR is a duplicate of Gitea PR 7)

### CODE REVIEW

File: Entities/Characters/Scripts/RunnerDefault.as [hunk 1]
```diff
@@ -58,12 +58,6 @@ void onAttach(CBlob@ this, CBlob@ attached, AttachmentPoint @attachedPoint)
 	this.server_DetachFrom( attached ); CRASHES*/
 }
 
-// set the Z back
```
Comment (AsuMagic) [inline Entities/Characters/Scripts/RunnerDefault.as:61]: Why is this removed?

This appears to cause an issue when leaving a catapult, for instance. The player Z appears to be modified when entering a catapult to visually appear behind it, but with this change, the player will still appear behind the catapult.

File: Entities/Materials/Ammunition/MaterialWaterArrow.as [hunk 2]
```diff
@@ -9,4 +9,17 @@ void onInit(CBlob@ this)
   this.maxQuantity = 2;
 
   this.getCurrentScript().runFlags |= Script::remove_after_this;
+
+  // Pickup: custom hover area
+  Vec2f[] hoverRect =
```
Comment (AsuMagic) [inline Entities/Materials/Ammunition/MaterialWaterArrow.as:14]: This is repeated across all arrow material scripts, maybe introduce a `genericArrowHoverRect` constant in a `ArrowCommon.as` script? Or some kind of constants script (e.g. `ArrowConstants.as`) though nothing else in KAG seems to follow such a convention (yet)...

File: Entities/Common/Controls/ImportantPickup.as [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  this.getSprite().SetZ(baseZ);
+  this.set_f32("important-pickup", baseZ);
+}
+
+void onAttach(CBlob@ this, CBlob@ attached, AttachmentPoint @attachedPoint)
+{
+  f32 baseZ = this.getSprite().getZ();
+  this.set_f32("important-pickup", baseZ);
```
Comment (AsuMagic) [inline Entities/Common/Controls/ImportantPickup.as:54]: Why do we need to check the sprite Z on attach? Looking at the logic my assumption would be that the sprite Z is not expected to change, so that `"important-pickup"` should always be correct.

This is not *really* important but is this something you've seen happen?

File: Entities/Common/Controls/ImportantPickup.as [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+void onDetach(CBlob@ this, CBlob@ detached, AttachmentPoint@ attachedPoint)
+{
+  f32 baseZ = this.get_f32("important-pickup");
+  this.getSprite().SetZ(baseZ);
+}
+
+// Add ImportantPickup.as to sprite_scripts in the .ctf file to enable rendering of 
+// the hover square (for debug purposes)
+void onRender(CSprite@ this)
```
Comment (AsuMagic) [inline Entities/Common/Controls/ImportantPickup.as:66]: The debug HUD is nice for testing... Though, TBH, the script may as well add itself to the sprite and have the function early exit if `g_debug == 0`. Because right now, if one wants to check/rework the hover shapes they would need to find all blobs that use the script to edit their `.cfg` individually.

File: Entities/Common/Controls/StandardControlsCommon.as [hunk 5]
```diff
@@ -40,3 +40,61 @@ bool isTapPickup(CBlob@ this, int ticks = 15)
...
+        Vec2f q = Vec2f(-polyVerts[i].x, polyVerts[i].y);
+        mirrored.push_back(q);
+      }
+
+      polygon = mirrored;
+    } 
+    else 
+    {
+      polygon = polyVerts;
+    }
```
Comment (AsuMagic) [inline Entities/Common/Controls/StandardControlsCommon.as:65]: Scrap the local variable. `Vec2f[] polyVerts` is already passed by value, so the original will be untouched anyway.

```suggestion
bool PointInsidePolygon( Vec2f Point, Vec2f[] polygon, Vec2f polyPos, bool facingLeft )
{
    // Mirror the polygon when the blob is facing left
    if (facingLeft) 
    {
      for ( int i = 0 ; i < polygon.length ; i++ )
      {
        polygon[i].x = -polygon[i].x;
      }
    }
```

File: Entities/Common/Controls/StandardControlsCommon.as [hunk 6]
```diff
@@ -40,3 +40,61 @@ bool isTapPickup(CBlob@ this, int ticks = 15)
...
+    {
+        return false;
+    }
+
+    bool inside = false;
+    for ( int i = 0, j = polygon.length - 1 ; i < polygon.length ; j = i++ )
+    {
+    	Vec2f pvi = polyPos + polygon[ i ];
+    	Vec2f pvj = polyPos + polygon[ j ];
+        if ( ( pvi.y > Point.y ) != ( pvj.y > Point.y ) &&
```
Comment (AsuMagic) [inline Entities/Common/Controls/StandardControlsCommon.as:91]: I'll trust that the math is correct as it appears to work :P

**Review by AsuMagic — CHANGES_REQUESTED** [review]
Does appear to work. Generally speaking the approach seems sound, but it will need a good amount of real world testing to find issues (bugs or finding other shapes that may need hover shapes). On top of my head, some things that should be good to check in testing:
- This changes a few Z values, most notably the keg one. It should be checked whether this causes unintended effects.
- We might as well look over some more defaults, such as the bomb Z which has not been mentioned in the PR.

File: Entities/Special/CTF/CTF_Flag.cfg [hunk 7]
```diff
@@ -3,7 +3,7 @@
 # @ array
 
 $sprite_factory                                   = generic_sprite
-@$sprite_scripts                                  = CTF_Flag.as;
+@$sprite_scripts                                  = CTF_Flag.as; ImportantPickup.as
```
[the remark this answers is not in the document]
Comment (Nananas) [reply Entities/Special/CTF/CTF_Flag.cfg:6]: No idea how I missed this, woops

File: Entities/Materials/Ammunition/MaterialBombArrow.as [hunk 8]
```diff
@@ -6,4 +6,19 @@ void onInit(CBlob@ this)
   this.set_u16('decay time', 300);
 
   this.getCurrentScript().runFlags |= Script::remove_after_this;
+
+  // Pickup: custom hover area
```
[the remark this answers is not in the document]
Comment (AsuMagic) [reply Entities/Materials/Ammunition/MaterialBombArrow.as:10]: None that really matters, no. It only sets `"decay time"` which is only matters to the server.

File: Entities/Special/CTF/CTF_Flag.cfg (hunk 7 above)
[the remark this answers is not in the document]
Comment (Nananas) [reply Entities/Special/CTF/CTF_Flag.cfg]: Removed

File: Entities/Materials/Ammunition/MaterialBombArrow.as (hunk 8 above)
[the remark this answers is not in the document]
Comment (Nananas) [reply Entities/Materials/Ammunition/MaterialBombArrow.as]: I removed the `SERVER_ONLY` define. I hope that's fine.

File: Entities/Characters/Scripts/RunnerDefault.as (hunk 1 above)
Comment (Nananas) [reply Entities/Characters/Scripts/RunnerDefault.as:61]: I reverted the remove. I always forget to test with catapults.

File: Entities/Materials/Ammunition/MaterialWaterArrow.as (hunk 2 above)
Comment (Nananas) [reply Entities/Materials/Ammunition/MaterialWaterArrow.as]: I moved the shape into `ArrowCommon.as`

File: Entities/Common/Controls/ImportantPickup.as (hunk 4 above)
Comment (Nananas) [reply Entities/Common/Controls/ImportantPickup.as:80]: Added a `g_debug` check

File: Entities/Common/Controls/StandardControlsCommon.as (hunk 6 above)
Comment (Nananas) [reply Entities/Common/Controls/StandardControlsCommon.as:80]: I got that from somewhere on the ~~ internet ~~

File: Entities/Common/Controls/ImportantPickup.as (hunk 3 above)
Comment (Nananas) [reply Entities/Common/Controls/ImportantPickup.as]: I removed `onAttach` for now. It does not make sense at the moment to modify the baseZ. Maybe just old code

**Review by AsuMagic — APPROVED** [review]

**Comment by AsuMagic** [comment]
This will need some real world testing and feedback, also keeping an eye out for anything weird with Z levels that may arise (notably related to attachments such as pickup, catapults, etc.).
